### ISSUE TITLE
Cargo doc explodes with ICE (rest of the tests seem to work?)

### ISSUE THREAD

**Issue by frankmcsherry** [rust-lang/rust#55690]
I think the following link should let you see what is going on. Relevant error text reproduced below. Reproduces locally too (`cargo test` passes; `cargo doc` does not).

https://travis-ci.org/frankmcsherry/differential-dataflow/builds/450756251

```
[snip]
 Documenting differential-dataflow v0.7.0 (/Users/mcsherry/Projects/differential-dataflow)                                                                    
error: internal compiler error: librustc/traits/select.rs:2536: Where clause `Binder(<_ as std::marker::Send>)` was applicable to `Obligation(predicate=Binder(TraitPredicate(<operators::ValueHistory<'_, _, _, _> as std::marker::Send>)),depth=2)` but now is not
                                                                                                                                                              
thread '<unnamed>' panicked at 'Box<Any>', librustc_errors/lib.rs:587:9                                                                                       
note: Run with `RUST_BACKTRACE=1` for a backtrace.                                                                                                            
                                                                                                                                                              
note: the compiler unexpectedly panicked. this is a bug.                                                                                                      
                                                                                                                                                              
note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports                                             
                                                                                                                                                              
note: rustc 1.30.0 (da5f414c2 2018-10-24) running on x86_64-apple-darwin                                                                                      
                                                                                                                                                              
error: Could not document `differential-dataflow`.                                                                                                            

Caused by:
  process didn't exit successfully: `rustdoc --crate-name differential_dataflow src/lib.rs --color always -o /Users/mcsherry/Projects/differential-dataflow/target/doc -L dependency=/Users/mcsherry/Projects/differential-dataflow/target/debug/deps --extern abomonation=/Users/mcsherry/Projects/differential-dataflow/target/debug/deps/libabomonation-970200789d96f2f0.rmeta --extern abomonation_derive=/Users/mcsherry/Projects/differential-dataflow/target/debug/deps/libabomonation_derive-ad4be47cd7451f35.dylib --extern fnv=/Users/mcsherry/Projects/differential-dataflow/target/debug/deps/libfnv-80a0fee956aec5d9.rmeta --extern serde=/Users/mcsherry/Projects/differential-dataflow/target/debug/deps/libserde-8336f27bf07c4b60.rmeta --extern serde_derive=/Users/mcsherry/Projects/differential-dataflow/target/debug/deps/libserde_derive-caf712db51cb0e10.dylib --extern timely=/Users/mcsherry/Projects/differential-dataflow/target/debug/deps/libtimely-f1aa95e8d5ae4deb.rmeta --extern timely_sort=/Users/mcsherry/Projects/differential-dataflow/target/debug/deps/libtimely_sort-58aa374f36467bac.rmeta` (exit code: 1)
```

**Comment by Aaron1011** [comment]
This appears to be fixed in Nightly as of `rustdoc 1.32.0-nightly (13dab66a6 2018-11-05)`

**Comment by QuietMisdreavus** [comment]
Can confirm it works in 1.29.1, so i bet it's related to https://github.com/rust-lang/rust/issues/52873 and its fix, https://github.com/rust-lang/rust/pull/54199.

**Comment by QuietMisdreavus** [comment]
...however, this one *doesn't* work on beta, so that PR isn't what fixed it. :confused:

**Comment by QuietMisdreavus** [comment]
Bisecting: Doesn't work on `rustc 1.31.0-nightly (4bd4e4130 2018-10-25)`, does work on `rustc 1.31.0-nightly (3e6f30ec3 2018-10-26)`, so i'm pretty sure that it was https://github.com/rust-lang/rust/pull/55258 that fixed it.

**Comment by QuietMisdreavus** [comment]
Bisection complete, https://github.com/rust-lang/rust/pull/55258 was the fix, i'm nominating for backport.

**Comment by pnkfelix** [comment]
triage: Tagging as P-high. But this was also fixed by PR #55258 which was approved for backport. So also closing in addition to tagging as P-high.
